starlark/int.go

```diff
@@ -26,7 +26,15 @@ func MakeInt64(x int64) Int {
 		}
 		return Int{&smallint[x]}
 	}
-	return Int{new(big.Int).SetInt64(x)}
+	if 0 <= x && int64(int32(x)) == x {
+		// x is guaranteed to fit into a single big.Word.
+		// Most starlark ints remain small,
+		// but math/big assumes that since you've chosen to use math/big,
+		// your ints will probably grow, so it overallocates.
+		// Avoid that overallocation by manually constructing a single-word slice.
+		return Int{new(big.Int).SetBits([]big.Word{big.Word(x)})}
```

Comment:
A further space optimization would be to allocate [1]big.Word inside the Int itself to avoid the slice allocation here.

Reply:
Sadly, that causes Int to no longer be pointer-shaped, so putting an Int into an interface (Value) allocates, and everything gets much worse.

We could potentially do this optimization inside math/big, but it is pretty invasive.

A deeper fix might be to have an Int type and a SmallInt type, both of which are Values, and migrate between them seamlessly as needed. I don't understand enough about the implementation yet to know how terrible an idea that is.

Reply:
> bottom 15 bits

That sounds hairy. Isn't the type word an *rtype pointer? What are the spare 15 bits?


> By the way, do you have good samples of realistic, performance sensitive Starlark code lying around?

No, and it's a problem. To validate performance changes I import the patched interpreter to a fairly complex application I'm developing and run it on a large real program; it's neither quick or convenient. I will keep my mind open for ideas for a good benchmark.

Reply:
> That sounds hairy. Isn't the type word an *rtype pointer? What are the spare 15 bits?

Sorry, I meant 15 available values; 4 bits. (One value has to be reserved.) In any case, yes, hairy and almost certainly not a good idea.

Now that Austin is working on ABI flexibility, though, three-word interfaces isn't quite as impossible any more.

> I will keep my mind open for ideas for a good benchmark.

Great.